internal/api/sms_provider/huawei_cloud.go

```diff
@@ -0,0 +1,259 @@
+	return fmt.Sprintf("%s Access=%s, SignedHeaders=%s, Signature=%s", SignAlgorithm, accessKeyStr, strings.Join(signedHeaders, ";"), signatureStr)
+}
+
+// Signature HWS meta
+type Signer struct {
+	Key    string
+	Secret string
+}
+
+// SignRequest set Authorization header
+func (s *Signer) Sign(request *http.Request) error {
+	var t time.Time
+	var err error
+	var date string
+	if date = request.Header.Get(HeaderXDateTime); date != "" {
+		t, err = time.Parse(DateFormat, date)
+	}
+	if err != nil || date == "" {
+		t = time.Now()
+		request.Header.Set(HeaderXDateTime, t.UTC().Format(DateFormat))
+	}
+	signedHeaders := SignedHeaders(request)
+	canonicalRequest, err := CanonicalRequest(request, signedHeaders)
+	if err != nil {
+		return err
+	}
+	stringToSignStr, err := StringToSign(canonicalRequest, t)
+	if err != nil {
+		return err
+	}
+	signatureStr, err := SignStringToSign(stringToSignStr, []byte(s.Secret))
+	if err != nil {
+		return err
+	}
+	authValueStr := AuthHeaderValue(signatureStr, s.Key, signedHeaders)
+	request.Header.Set(HeaderXAuthorization, authValueStr)
+	return nil
+}
+
+type HuaweiCloudResponse struct {
+	Code        string       `json:"code"`
+	Description string       `json:"description"`
+	Result      []ResultItem `json:"result"`
+}
+
+type ResultItem struct {
+	Total      int    `json:"total"`
+	OriginTo   string `json:"originTo"`
+	CreateTime string `json:"createTime"`
+	From       string `json:"from"`
+	SmsMsgId   string `json:"smsMsgId"`
+	CountryId  string `json:"countryId"`
+	Status     string `json:"status"`
+}
+
+type HuaweiCloudProvider struct {
+	Config *conf.HuaweiCloudProviderConfiguration
+}
+
+// SendMessage implements SmsProvider.
+func (p *HuaweiCloudProvider) SendMessage(phone string, message string, channel string, otp string) (string, error) {
+	switch channel {
+	case SMSProvider:
+		return p.SendSms(phone, message, otp)
+	default:
+		return "", fmt.Errorf("channel type %q is not supported for Messagebird", channel)
+	}
+}
+
+func (p *HuaweiCloudProvider) SendSms(phone, templateId, otp string) (string, error) {
+	params := buildRequestParams(p.Config.ChannelNumber, phone, templateId, otp, p.Config.ChannelName)
+	url := p.Config.ApiPath + "/sms/batchSendSms/v1"
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer([]byte(params)))
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	s := Signer{
+		Key:    p.Config.ApiKey,
+		Secret: p.Config.ApiSecret,
+	}
+	serr := s.Sign(req)
+	if serr != nil {
+		return "", serr
+	}
+
+	client := &http.Client{Timeout: defaultTimeout}
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	r := &HuaweiCloudResponse{}
+	derr := json.NewDecoder(resp.Body).Decode(r)
+	if derr != nil {
+		return "", derr
+	}
+
+	if len(r.Result) == 0 {
+		return "", errors.New("result is empty")
+	}
```

Comment:
please resolve the `code` in response body

ref: https://support.huaweicloud.com/api-msgsms/sms_05_0001.html#section8

Reply:
when create the error, please also carry the field `code` and field `description`